datasets/flwr_datasets/partitioner/natural_id_partitioner.py

```diff
@@ -42,6 +47,61 @@ def _create_int_partition_id_to_natural_id(self) -> None:
             zip(range(len(unique_natural_ids)), unique_natural_ids)
         )
 
+    def _create_natural_id_to_int_partition_id(self) -> None:
+        """Create a mapping from unique client ids from dataset to int indices.
+
+        Natural ids come from the column specified in `partition_by`. This object is
+        inverse of the `self._partition_id_to_natural_id`. This method assumes that
+        `self._partition_id_to_natural_id` already exist.
+        """
+        self._natural_id_to_partition_id = {
+            value: key for key, value in self._partition_id_to_natural_id.items()
+        }
+
+    def _create_partition_id_to_indices(self) -> None:
+        """Create an assignment of indices to the partition indices."""
+        natural_ids = np.array(self.dataset[self._partition_by])
+        unique_natural_ids = self.dataset.unique(self._partition_by)
+
+        none_present = False
+        if None in unique_natural_ids:
+            none_present = True
+            dtype = self.dataset.features[self._partition_by].dtype
+            none_replacement: Union[int, str]
+            if dtype == "string":
+                none_replacement = "None"
+                # Ensure the replacement is not in the dataset
+                while True:
+                    if none_replacement not in unique_natural_ids:
+                        break
+                    none_replacement += "1"
```

Comment:
This will result in a `none_replacement` in the form `"None1111111111...."`. Should we instead do the `+= 1` and only when ending the `if` statement convert it to `str`? -- I understand this is just for internal use in the scope of this method though.

Reply:
I need to change it into the string in each loop to check if it's not present in the `unique_natural_ids` but I can append based on counter so instead of `None1111` (if None1 None2 None3 strings were in the partition_by column) we might have `None4` (if None1, None2, None3 were in partition_by column). 
Both of them are very unlikely. This is to protect ourselves from an unlikely situation that the column of `partition_by` contains "None" (str) and None (None type). And then theoretically we might go further (maybe "None1" string is in the column so it can be changed further.

So, I'll change it to the counter.